adapters/rack/lib/opentelemetry/adapters/rack/middlewares/tracer_middleware.rb

```diff
@@ -0,0 +1,253 @@
+          def request_span_attributes(env:)
+            {
+              'component' => 'http',
+              'http.method' => env['REQUEST_METHOD'],
+              'http.host' => env['HOST'],
+              'http.scheme' => env['rack.url_scheme'],
+              'http.target' => fullpath(env)
+            }.merge(allowed_request_headers(env))
+          end
+
+          # e.g., "/webshop/articles/4?s=1":
+          def fullpath(env)
+            query_string = env['QUERY_STRING']
+            path = env['SCRIPT_NAME'] + env['PATH_INFO']
+
+            query_string.empty? ? path : "#{path}?#{query_string}"
+          end
+
+          # https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-http.md#name
+          #
+          # recommendation: span.name(s) should be low-cardinality (e.g.,
+          # strip off query param value, keep param name)
+          #
+          # see http://github.com/open-telemetry/opentelemetry-specification/pull/416/files
+          def create_request_span_name(request_uri_or_path_info)
+            # NOTE: dd-trace-rb has implemented 'quantization' (which lowers url cardinality)
+            #       see Datadog::Quantization::HTTP.url
+
+            if (implementation = config[:url_quantization])
+              implementation.call(request_uri_or_path_info)
+            else
+              request_uri_or_path_info
+            end
+          end
+
+          # catch exceptions that may be raised in the middleware chain
+          # Note: if a middleware catches an Exception without re raising,
+          # the Exception cannot be recorded here.
+          def record_and_reraise_error(error, request_span:)
+            request_span.record_error(error)
+            request_span.status = OpenTelemetry::Trace::Status.new(
+              OpenTelemetry::Trace::Status::UNKNOWN_ERROR,
+              description: "Unhandled exception of type: #{error.class}"
+            )
+
+            raise error
+          end
+
+          def set_attributes_after_request(span, status, headers, _response)
+            span.status = OpenTelemetry::Trace::Status.http_to_status(status)
+            span.set_attribute('http.status_code', status)
+
+            # NOTE: if data is available, it would be good to do this:
+            # set_attribute('http.route', ...
+            # e.g., "/users/:userID?
+            span.set_attribute('http.status_text', ::Rack::Utils::HTTP_STATUS_CODES[status])
+
+            allowed_response_headers(headers).each { |k, v| span.set_attribute(k, v) }
+          end
+
+          # @return Hash
+          def allowed_request_headers(env)
+            return EMPTY_HASH if self.class.allowed_rack_request_headers.empty?
+
+            {}.tap do |result|
+              self.class.allowed_rack_request_headers.each do |key, value|
+                result[value] = env[key] if env.key?(key)
+              end
+            end
+          end
+
+          # @return Hash
```

Comment:
I'm not sure we need to annotate this return. The rest of the methods do not have `@return` anyhow.